File: VB/DXApplication/Form1.vb
Imports DevExpress.Xpo
Imports DevExpress.Xpo.DB
Imports System
Imports System.Data
Imports System.IO
Imports System.Linq

Namespace DXApplication
	Partial Public Class Form1
		Inherits DevExpress.XtraEditors.XtraForm

		Public Sub New()
			InitializeComponent()
			SetupLookupDataSources()
			DataSourceTypeComboBox.SelectedIndex = 0
		End Sub

        Private _dataLayer As IDataLayer
        Public ReadOnly Property DataLayer() As IDataLayer
			Get
                If _dataLayer Is Nothing Then
                    Dim dataStore As New InMemoryDataStore()
                    dataStore.ReadXml(Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "data.xml"))
                    _dataLayer = New SimpleDataLayer(dataStore)
                    Return _dataLayer
                End If
                Return _dataLayer
            End Get
		End Property

        Private _unitOfWork As UnitOfWork
        Public ReadOnly Property UnitOfWork() As UnitOfWork
			Get
                If _unitOfWork Is Nothing Then
                    _unitOfWork = New UnitOfWork(DataLayer)
                    _unitOfWork.TrackPropertiesModifications = True
                End If
                Return _unitOfWork
            End Get
		End Property

		Private Sub SetupMainDataSource(ByVal dataSourceType As String)
			Dim oldDataSource As IDisposable = TryCast(ProductListSource.DataSource, IDisposable)
			If oldDataSource IsNot Nothing Then
				oldDataSource.Dispose()
			End If
			Select Case dataSourceType
				Case "XPCollection"
					ProductListSource.DataSource = New XPCollection(UnitOfWork, GetType(Products))
				Case "XPServerCollectionSource"
					Dim ds As New XPServerCollectionSource(UnitOfWork, GetType(Products))
					ds.AllowEdit = True
					ds.TrackChanges = True
					ProductListSource.DataSource = ds
				Case "List of Objects"
                    ProductListSource.DataSource = UnitOfWork.Query(Of Products)().ToList()
                    ProductListSource.ObjectClassInfo = UnitOfWork.GetClassInfo(Of Products)()
                Case "Single Object"
                    ProductListSource.DataSource = UnitOfWork.Query(Of Products)().First()
                    ProductListSource.ObjectClassInfo = UnitOfWork.GetClassInfo(Of Products)()
            End Select
		End Sub

		Private Sub SetupLookupDataSources()
            CategoryListSource.DataSource = UnitOfWork.Query(Of Categories)().OrderBy(Function(t) t.CategoryName).ToList()
            CategoryListSource.Dictionary = UnitOfWork.Dictionary
            SupplierListSource.DataSource = New XPCollection(Of Suppliers)(UnitOfWork, Nothing, New SortProperty("CompanyName", SortingDirection.Ascending))
		End Sub

		Public Sub ReCreateUnitOfWork()
            _unitOfWork.Dispose()
            _unitOfWork = Nothing
            SetupLookupDataSources()
			SetupMainDataSource(DataSourceTypeComboBox.EditValue.ToString())
		End Sub

		Private Sub DataSourceTypeComboBox_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles DataSourceTypeComboBox.SelectedIndexChanged
			SetupMainDataSource(DataSourceTypeComboBox.EditValue.ToString())
		End Sub

		Private Sub simpleButton1_Click(ByVal sender As Object, ByVal e As EventArgs) Handles simpleButton1.Click
			ReCreateUnitOfWork()
		End Sub
	End Class
End Namespace

File: CS/DXApplication/Form1.cs
using DevExpress.Xpo;
using DevExpress.Xpo.DB;
using System;
using System.Data;
using System.IO;
using System.Linq;

namespace DXApplication
{
    public partial class Form1 : DevExpress.XtraEditors.XtraForm
    {
        public Form1()
        {
            InitializeComponent();
            SetupLookupDataSources();
            DataSourceTypeComboBox.SelectedIndex = 0;
        }

        IDataLayer dataLayer;
        public IDataLayer DataLayer
        {
            get
            {
                if (dataLayer == null)
                {
                    InMemoryDataStore dataStore = new InMemoryDataStore();
                    dataStore.ReadXml(Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "data.xml"));
                    dataLayer = new SimpleDataLayer(dataStore);
                    return dataLayer;
                }
                return dataLayer;
            }
        }

        UnitOfWork unitOfWork;
        public UnitOfWork UnitOfWork
        {
            get
            {
                if (unitOfWork == null)
                {
                    unitOfWork = new UnitOfWork(DataLayer);
                    unitOfWork.TrackPropertiesModifications = true;
                }
                return unitOfWork;
            }
        }

        void SetupMainDataSource(string dataSourceType)
        {
            IDisposable oldDataSource = ProductListSource.DataSource as IDisposable;
            if (oldDataSource != null)
            {
                oldDataSource.Dispose();
            }
            switch (dataSourceType)
            {
                case "XPCollection":
                    ProductListSource.DataSource = new XPCollection(UnitOfWork, typeof(Products));
                    break;
                case "List of Objects":
                    ProductListSource.DataSource = UnitOfWork.Query<Products>().ToList();
                    ProductListSource.ObjectClassInfo = UnitOfWork.GetClassInfo<Products>();
                    break;
                case "Single Object":
                    ProductListSource.DataSource = UnitOfWork.Query<Products>().First();
                    ProductListSource.ObjectClassInfo = UnitOfWork.GetClassInfo<Products>();
                    break;
            }
        }

        void SetupLookupDataSources()
        {
            CategoryListSource.DataSource = UnitOfWork.Query<Categories>().OrderBy(t => t.CategoryName).ToList();
            CategoryListSource.Dictionary = UnitOfWork.Dictionary;
            SupplierListSource.DataSource = new XPCollection<Suppliers>(UnitOfWork, null, new SortProperty("CompanyName", SortingDirection.Ascending));
        }

        public void ReCreateUnitOfWork()
        {
            unitOfWork.Dispose();
            unitOfWork = null;
            SetupLookupDataSources();
            SetupMainDataSource(DataSourceTypeComboBox.EditValue.ToString());
        }

        private void DataSourceTypeComboBox_SelectedIndexChanged(object sender, EventArgs e)
        {
            SetupMainDataSource(DataSourceTypeComboBox.EditValue.ToString());
        }

        private void simpleButton1_Click(object sender, EventArgs e)
        {
            ReCreateUnitOfWork();
        }
    }
}
